### DISCUSSION TITLE
remote read of NWB files using MATLAB

### DISCUSSION THREAD

**Discussion by bendichter** [dandi/helpdesk#59] [Q&A] [answered]
It is supposed to be possible to do a remote read of s3 hdf5 files in MATLAB, but when I try to get this:

```matlab
>> H5F.open('https://dandiarchive.s3.amazonaws.com/blobs/58c/537/58c53789-eec4-4080-ad3b-207cf2a1cac9')

Error using hdf5lib2
Unable to access
'https://dandiarchive.s3.amazonaws.com/blobs/58c/537/58c53789-eec4-4080-ad3b-207cf2a1cac9'.
The specified URL scheme is invalid.
```

I am guessing that "The specified URL scheme is invalid." is referring to the fact that the url does not start with "s3://". Why doesn't it? Do we have an alternative url that fits this pattern?

**Comment by satra** [comment]
this is because the `hdf5 ros3` driver uses `libcurl` underneath which only understands the `http` protocol.  

you can replace: `https://dandiarchive.s3.amazonaws.com/` with `s3://dandiarchive` to get the s3 location. i believe i've tried this in the past with matlab, and i'm not sure that will work either. i don't have matlab on my current machine to try. 

here is a bit of code:

```python
    s3loc = [val for val in contenturl if ".s3.amazonaws" in val]
    assert len(s3loc) == 1
    bucket = s3loc[0].split("//")[1].split(".s3.amazonaws")[0]
```

**Reply by bendichter** [reply]
@satra 

So it should be "s3://dandiarchive/blobs/58c/537/58c53789-eec4-4080-ad3b-207cf2a1cac9"? I have tried a few variations of this but cannot get it to read in `h5py`.

```python
file = File("s3://dandiarchive/blobs/58c/537/58c53789-eec4-4080-ad3b-207cf2a1cac9", "r", driver="ros3")
```

```python
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
/tmp/ipykernel_413/1841186545.py in <module>
----> 1 file = File("s3://dandiarchive/blobs/58c/537/58c53789-eec4-4080-ad3b-207cf2a1cac9", "r", driver="ros3")
      2 file.keys()

/opt/conda/lib/python3.9/site-packages/h5py/_hl/files.py in __init__(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, fs_page_size, page_buf_size, min_meta_keep, min_raw_keep, locking, **kwds)
    505                                  fs_persist=fs_persist, fs_threshold=fs_threshold,
    506                                  fs_page_size=fs_page_size)
--> 507                 fid = make_fid(name, mode, userblock_size, fapl, fcpl, swmr=swmr)
    508 
    509             if isinstance(libver, tuple):

/opt/conda/lib/python3.9/site-packages/h5py/_hl/files.py in make_fid(name, mode, userblock_size, fapl, fcpl, swmr)
    218         if swmr and swmr_support:
    219             flags |= h5f.ACC_SWMR_READ
--> 220         fid = h5f.open(name, flags, fapl=fapl)
    221     elif mode == 'r+':
    222         fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/h5f.pyx in h5py.h5f.open()

OSError: Unable to open file (invalid SCHEME construction)
```

**Reply by satra** [reply]
h5py uses the hdf5 ros3 driver which only understands the http protocol. you cannot use the s3 protocol with h5py. i don't think matlab enables the ros3 vfd flag. their s3 access is a completely different route.

if you want to use the s3 protocol in python, you will have to use s3fs and xarray.

```python
import xarray as xr
import hdf5plugin
import s3fs

s3_url = "s3://dandiarchive/blobs/e08/02c/e0802c3e-5492-42a6-960a-bcf5b3cfd239"
fs = s3fs.S3FileSystem(anon=True)
with fs.open(s3_url, "rb") as fp:
    ds = xr.open_dataset(fp, engine="h5netcdf", phony_dims='access')
    data2 = ds["6"].data
```

**Reply by bendichter** [reply]
@satra 

Actually, you can:

```python
import hdf5plugin
import s3fs
import h5py

s3_url = "s3://dandiarchive/blobs/e08/02c/e0802c3e-5492-42a6-960a-bcf5b3cfd239"
fs = s3fs.S3FileSystem(anon=True)
with fs.open(s3_url, "rb") as fp:
    file = h5py.File(fp)
    print(file.keys())
```

but I'm still not sure how to translate this to MATLAB

**Reply by satra** [reply]
sure s3fs is doing the work for you. there is a similar setup in matlab, but i think there is no `anon=True` option. so you need to setup access keys or ask mathworks to support anonymous access like boto does.

**Reply by nvishnoiMW** [reply]
Hi Ben, 

I can successfully use the following in MATLAB: 

>> s3_url = "s3://dandiarchive/blobs/e08/02c/e0802c3e-5492-42a6-960a-bcf5b3cfd239"

s3_url = 

    "s3://dandiarchive/blobs/e08/02c/e0802c3e-5492-42a6-960a-bcf5b3cfd239"
 
>> H5F.open(s3_url)


ans = 

   72057594037927936

**Reply by bendichter** [reply]
@nvishnoiMW thanks so much for the help. When I run this I get

```matlab
s3_url = "s3://dandiarchive/blobs/e08/02c/e0802c3e-5492-42a6-960a-bcf5b3cfd239"
H5F.open(s3_url)

Error using hdf5lib2
Unable to open or create
's3://dandiarchive/blobs/e08/02c/e0802c3e-5492-42a6-960a-bcf5b3cfd239'
due to internal error.

Error in H5F.open (line 130)
file_id = H5ML.hdf5lib2('H5Fopen', filename, flags, fapl, is_remote);
```

I am using MATLAB 2021b on OSX. Is the s3 feature only available in a more recent release or another operating system?

**Reply by nvishnoiMW** [reply]
@bendichter I tried MATLAB R2021b on Mac and I was able to open the file on s3 using H5F.open. 
Couple of things to try: 
a. if you have access to a different machine with same MATLAB release, can you try the steps there?
b. On your machine, what do you get if you run:
>> fopen (s3_url)

**Reply by bendichter** [reply]
```matlab
>> fopen(s3_url)

ans =

    -1
```

Maybe something is broken in my system. I'm asking others to test this out for me. If it works for them I'll look into reinstalling MATLAB

**Reply by bendichter** [reply]
@nvishnoiMW, @lawrence-mbf just tried on 2021b Windows and got the exact same results as me for `H5F.open` and `fopen`.

**Reply by nvishnoiMW** [reply]
@bendichter One of my colleagues was able to reproduce the issue and we noticed they had their S3 credentials set to a specific value. Can you please try the following before accessing the aforementioned s3 url?

setenv('AWS_ACCESS_KEY_ID', ''); 
setenv('AWS_SECRET_ACCESS_KEY', '');

Hope this resolves the issue. 
Thanks,
Nalini

**Reply by bendichter** [reply]
Did you want me to actually set those to empty strings or to real ID and KEY values?

**Reply by bendichter** [reply]
For empty strings the problem persists.

**Reply by nvishnoiMW** [reply]
Since it is a publicly accessible bucket we want to reset them. 
>> getenv('AWS_ACCESS_KEY_ID')

ans =

  0×0 empty char array

>> getenv('AWS_SECRET_ACCESS_KEY')

ans =

  0×0 empty char array

**Reply by bendichter** [reply]
```matlab
>> getenv('AWS_ACCESS_KEY_ID')


ans =

  0×0 empty char array

>> 
>> getenv('AWS_SECRET_ACCESS_KEY')


ans =

  0×0 empty char array

>> 
>> H5F.open(s3_url)
Error using hdf5lib2
Unable to open or create
's3://dandiarchive/blobs/e08/02c/e0802c3e-5492-42a6-960a-bcf5b3cfd239'
due to internal error.

Error in H5F.open (line 130)
file_id = H5ML.hdf5lib2('H5Fopen', filename, flags, fapl, is_remote);
```

**Reply by nvishnoiMW** [reply]
is it same for @lawrence-mbf?

**Reply by nvishnoiMW** [reply]
@bendichter I am assuming the issue persists with fopen as well even after resetting the variables?

**Reply by lawrence-mbf** [reply]
@nvishnoiMW
![fopen](https://user-images.githubusercontent.com/33462079/156812640-702cb510-7879-44ce-af00-9a1681d316d0.png)

As mentioned above, this is on Windows 10 using MATLAB R2021b.

**Comment by bendichter** [comment, marked as answer]
The solution involved 2 fixes. 

First to use this url: s3_url = "s3://dandiarchive/blobs/e08/02c/e0802c3e-5492-42a6-960a-bcf5b3cfd239"

Second, to fix a credentials issue, @nvishnoiMW proposed in another thread to delete stale credentials by deleting the file `~/.aws/credentials` (on Windows something like `C:/Users/username/.aws/credentials`)

**Comment by bendichter** [comment]
@satra could you help me figure out what's going on here?

```python
from h5py import File

s3_url = "https://dandiarchive.s3.amazonaws.com/blobs/58c/537/58c53789-eec4-4080-ad3b-207cf2a1cac9"

file = File(s3_url, mode = "r", driver="ros3")
file.attrs['nwb_version']
```

```
2.2.5
```

```python
bucket = s3_url.replace("https://dandiarchive.s3.amazonaws.com", "s3://dandiarchive")
print(bucket)
```

```
s3://dandiarchive/blobs/58c/537/58c53789-eec4-4080-ad3b-207cf2a1cac9
```

```python
file = File(bucket, mode = "r", driver="ros3")
```

```python
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
Input In [28], in <module>
      9 bucket = s3_url.replace("https://dandiarchive.s3.amazonaws.com/", "s3://dandiarchive")
     10 print(bucket)
---> 11 file = File(bucket, mode = "r", driver="ros3")

File /opt/conda/lib/python3.9/site-packages/h5py/_hl/files.py:507, in File.__init__(self, name, mode, driver, libver, userblock_size, swmr, rdcc_nslots, rdcc_nbytes, rdcc_w0, track_order, fs_strategy, fs_persist, fs_threshold, fs_page_size, page_buf_size, min_meta_keep, min_raw_keep, locking, **kwds)
    502     fapl = make_fapl(driver, libver, rdcc_nslots, rdcc_nbytes, rdcc_w0,
    503                      locking, page_buf_size, min_meta_keep, min_raw_keep, **kwds)
    504     fcpl = make_fcpl(track_order=track_order, fs_strategy=fs_strategy,
    505                      fs_persist=fs_persist, fs_threshold=fs_threshold,
    506                      fs_page_size=fs_page_size)
--> 507     fid = make_fid(name, mode, userblock_size, fapl, fcpl, swmr=swmr)
    509 if isinstance(libver, tuple):
    510     self._libver = libver

File /opt/conda/lib/python3.9/site-packages/h5py/_hl/files.py:220, in make_fid(name, mode, userblock_size, fapl, fcpl, swmr)
    218     if swmr and swmr_support:
    219         flags |= h5f.ACC_SWMR_READ
--> 220     fid = h5f.open(name, flags, fapl=fapl)
    221 elif mode == 'r+':
    222     fid = h5f.open(name, h5f.ACC_RDWR, fapl=fapl)

File h5py/_objects.pyx:54, in h5py._objects.with_phil.wrapper()

File h5py/_objects.pyx:55, in h5py._objects.with_phil.wrapper()

File h5py/h5f.pyx:106, in h5py.h5f.open()

OSError: Unable to open file (invalid SCHEME construction)
```

**Reply by satra** [reply]
@bendichter - as noted earlier in this thread `h5py` is based on `hdf5` which is based on `libcurl` and only accepts the `http` protocol, not the `s3` protocol.

**Reply by bendichter** [reply]
ah, ok, I understood that MATLAB couldn't take the https, but not that h5py couldn't take s3
